### ISSUE TITLE
BLE connection problems in v6

### ISSUE THREAD

**Issue by emilohman** [athombv/apps-sdk-issues#203]
I'm getting a lot of error reports for my app https://homey.app/a/se.emilohman.plejd from users and their problem is that they can't connect to their devices over BLE after upgrading to Homey firmware v6. The error is just a timeout error when connecting to the advertisement.

It's not affecting all users, it's working for me. Do you experience similar problems with BLE connections?

**Comment by mrlukasbos** [comment]
Hi, thank you for reporting. Connection Timeout issues could appear for multiple reasons. 

- It may occur when it is tried to connect with multiple devices at once. (this is due to queuing)
- It may occur when the BLE devices are at the edge of the range. (Or when there is much interference)
- It could be an issue related to some specific behaviour of Plejd devices. If there is a smartphone app for example that also connects to this device, it may prevent Homey from connecting.

Could you perhaps share some more details on the code you use for connecting to BLE devices?

**Comment by emilohman** [comment]
Thanks for your reply!
It's just that it was working before v6 and now it's stopped working for a lot of users.
Absolutely, the code is available here: https://github.com/emilohman/homey-plejd/blob/master/drivers/plejd/driver.js
The `currentAdvertisement.__peripheral = null;` line is for some old bug in Homey but I will remove that.

I'm thinking of make a test where I try to connect to a different advertisements (they are all meshed together) from the sorted (by rssi) list when reconnecting.

**Comment by mrlukasbos** [comment]
Could be an issue with v6 indeed. Do you have an idea of the range of these devices? Does it occur more often when devices are further away?

**Comment by emilohman** [comment]
Unfortunately, I don't know. Just that it worked in v5 but not in v6. Is it any changes in v6 that can effect the range for BLE?

**Comment by mrlukasbos** [comment]
There are a few differences that may have an impact:

- Performing a discovery scan in V6 is limited to 5s instead of 30s, maybe the advertisements get out of date and a new advertisement is never found. 

- V5 had a bug, which caused the issue that `homey.ble.disconnect()` would not work. This has now been fixed so you disconnect() functions now actually have an effect. Maybe this causes more reconnections and triggers the issue more often.

**Comment by emilohman** [comment]
I can see that users that have 10-15 devices added just find 1 or 2 on discovery.
In the sdk it says 5s is default but max is 30s i that correct? https://apps-sdk-v3.developer.athom.com/tutorial-Bluetooth%20Low%20Energy.html

**Comment by mrlukasbos** [comment]
No. The documentation is outdated. Currently the max has been limited to 5s to prevent another issue, but I am looking for a way to put back this timeout. It may definitely have something to do with your issue. 

Over the upcoming weeks we may be tweaking BLE a bit in the experimental channel.

**Comment by emilohman** [comment]
This is still a problem for my users in v7.. have you found a fix for the discovery?Is something else changed from v5? Something that can affect the signal strength.

**Comment by mrlukasbos** [comment]
There was indeed a problem with discovery which I have recently resolved (V7.0.0). Apart from the shorter discovery time (which should not make a large difference) there is no known issue that could prevent Homey from connecting to the peripherals - there were some issues with this which have been fixed specifically in V6.0.0. The BLE range has not been affected, although the BLE range always has been rather short (max 10m). 

Although I don't have any Plejd ligths, I cannot reproduce your issue with many other BLE devices. Could it be something happening on the Plejd-side? For example users that remove the power from the Plejds at some point?

**Comment by emilohman** [comment]
Ok.. it's just that it worked in v5 and stopped working in v6. I cannot reproduce my self, but there is user that contact me with problem every week. 
Nothing seems to have changed on the Plejd side. Plejd lights is a dimmer puck that is located in the wall (like fibaro and many other) so the power is only remove with the power fuse.

What happens is that my app finds 1-2 devices of about 30 devices... most often it managed to connect.. but after a little while it gets a timeout and starts to reconnect..

**Comment by denniedegroot** [comment]
@mrlukasbos I can get Homey in a state where the 'disconnect' trigger is immediatly triggered after a connect. This happens after I get a "Error: BLE Timeout". Restarting my app (https://homey.app/nl-nl/app/com.luke-roberts/Luke-Roberts/) doesn't work. I have to restart Homey in order for the app/device to work again.

```
this._peripheral = await this._device.connect();
this._peripheral.once('disconnect', () => {}); // This triggers immediately once a "BLE Timeout has occured".
```

**Comment by mrlukasbos** [comment]
Does the 'disconnect' event trigger immediately, or does it take a few seconds (15 or more)? 

When a connect-attempt takes too long a timeout error is thrown and Homey will disconnect the peripheral (even while it is not connected) to clear state.

**Comment by denniedegroot** [comment]
When it is in this 'error state' it is called immediately (within a second) every time.
In a normal working state the disconnect is called after 6 seconds by the device or is executed manually.

Edit: Even if I remove this disconnect trigger it will never connect again and always gives a "BLE Timeout" until Homey is restarted.

**Comment by ErikHenefalk** [comment]
@mrlukasbos I am using @emilohman s Plejd-app in Homey. The app worked very well with Homey ver. 5, but with ver. 6 and all versions after the app stopped working. Only if I restart Homey and the Plejd-app it works for about 30 seconds and after that the devices (all 27) stop responding and I get a red message saying "Timed out after 30000 ms". Emil has been kind and tried to help me, but the error seems to be in Homey and not his Plejd-app. Is it possible for me to send you some kind of diagnostics-report so that you can see what happens in my Homey?

**Comment by mojento** [comment]
@mrlukasbos I have issues to connect my newly installed Plejd-devices to homey via @emilohman Homey app 'Plejd'.
I have 9 of them installed and added in Plejd-app in Homey since v5. For some reason I cannot add the Plejd BLE-devices anymore. They are discoverably for 30 sek. and then they disappear.

**Comment by mrlukasbos** [comment]
Thank you all for responding, and please excuse me for taking such a long time to respond. 

@denniedegroot Seems like a cache issue and might be related to the Plejd issue as well. I'll take a look.

 @ErikHenefalk A diagnostic report would be great to have. You can create one in the Homey Mobile app. You can create it under Settings -> General -> Create Diagnostics Report. You will receive a code which I can use to access the report.

@mojento This issue has indeed been reported by several other users as well. Unfortunately I can't replicate it at this moment. If you could send a diagnostics report as well then maybe that can help me trace the issue.

**Comment by ErikHenefalk** [comment]
@mrlukasbos The code to my diagnostics report is : E54082

**Comment by mrlukasbos** [comment]
Thanks for the diagnostics report. Everything seems to work as expected, however the Plejd app seems to be hitting some internal restrictions that have been introduced in v6.0.0 to improve the stability of the BLE chip. Perhaps I can loosen those restrictions a bit, as many of you state that the Plejd app previously worked fine.

If more people can send diagnostic reports that would help me very much. 

@emilohman You wrote that you can't reproduce the issue. Could you perhaps send me a diagnostic report as well, such that I have a "healthy" log which I can use as comparison?

**Comment by emilohman** [comment]
@mrlukasbos Here it is: A8525C

Sounds great that you have found something! Is there anything I can change in my code to prevent those restrictions?

**Comment by mrlukasbos** [comment]
In your log there is no communication with any Plejd device visible. Perhaps you can restart your app and show a diagnostic report right after that?

In the logs I see that there is a lot of BLE communication going on, especially many calls using writeCharacteric() and it takes quite some time for all those calls to complete. That may eventually start causing timeouts. However, that is not necessarily an error in your app; Homey should be able to support that. 

I hope to receive some more diagnostic reports, perhaps I can find a pattern.

**Comment by emilohman** [comment]
Here is a new report: 09FC26
Where I have restarted the app and toggled one light on/off.

**Comment by ErikHenefalk** [comment]
@mrlukasbos and @emilohman
Hi, I have now updated the Plejd-app to v 3.2.5. I have restarted Homey and the Plejd-app twice and unfortunately there is no difference from before the update.
Here is the diagnostics-report after the update and restarts: F37E33
Regards, Erik

**Comment by emilohman** [comment]
@mrlukasbos Any news on this? Anything we can help with?

**Comment by emilohman** [comment]
And then there was silence...

**Comment by mrlukasbos** [comment]
I hope to receive more logs (from different users) to help me further scope down the issue. For now that would help a lot. 

Until then I cannot do much. I do find it strange that the issue appears for some users while for other users it works fine, so I hope to find a critical difference somewhere.

**Comment by ErikHenefalk** [comment]
@emilohman, you said earlier in the thread that there had been a lot of other people contacting you about this problem. Would you be able to contact them and do you think they could send diagnostics reports as well?

@mrlukasbos, One thing I noticed is that Homey detects that the Plejd-devices have switched state (on/off), but only after sometime between 3 – 10 minutes. I have a flow that is triggering when one of the Plejd-devices changes between on/off, and that flow is triggered some time between 3-10 minutes after the device is either turned on or off. Here is a diagnostics-report when that happens: 87AC63

Could it be some hardware-failure? My Homey-model is: Homey (Early 2016). Which model of Homey do you have @emilohman, since the app works for you?

Are there other apps in Homey that are handling Bluetooth-devices that have similar problems?

**Comment by mrlukasbos** [comment]
Homey 7.1.5 is available in which I did some changes for BLE. Perhaps this improves things. Could you send me your experiences with the new Homey version and send new diagnostics?

**Comment by ErikHenefalk** [comment]
Hi @mrlukasbos 
Here is the new diagnostics report 9BE9C5
The Plejd app is working a little better now. After a restart of Homey after the update to 7.1.5 and after another restart of the Plejd app I could control the Plejd system for about 40 seconds before it locked timed out like before. The new thing is that Plejd devices started to work again after some minutes. They then stopped working again and now don't work at all.

**Comment by ErikHenefalk** [comment]
@mrlukasbos
Here is another diagnostics report, now after another restart of Homey and deaktivation/aktivation of the Plejd-app.
759DA0
The Plejd-app works for about 40 seconds before they stop working and the "timed out" notification comes up, then it looks like all Plejd-devices are deactivated and then reactivated again. The Plejd-devices then works for a few seconds and then stops working again.

**Comment by ErikHenefalk** [comment]
Hi again @mrlukasbos
I am starting to doubt that there will ever be a software solution to this bluetooth-problem. Don't you also think this is some kind of hardware-failure? If it is a hardware-failure, is it then possible to fix it? Can I send my Homey to Athom for a repair?
Regards,
Erik

**Comment by mrlukasbos** [comment]
Hi Erik, 

I am certain that this is not a hardware failure. A new version is currently in the works which applies some further improvements. It may take a few weeks before the new version is available.
I am sorry for all the inconvenience.

kind regards,
Lukas

**Comment by mrlukasbos** [comment]
@ErikHenefalk could you perhaps update your Homey to v7.2.1-rc.2? I applied some BLE fixes.

**Comment by ErikHenefalk** [comment]
@mrlukasbos I haven't activated the downloads of experimental updates on my Homey. Is it the "Power user" under the "Experiment"-menu I have to activate to get them? If I activate the experimental updates, can I then go back to only downloading non-experimental updates again after that?

**Comment by Dijker** [comment]
You can switch to Experimental updates here:
https://tools.developer.homey.app/tools/settings
Don't read the old note there, that is not updated from FW 2 or 3 ;-) 

once you are on a stable firmware version and disable the Experimental updates you will stay on stable updates from that moment.

**Comment by ErikHenefalk** [comment]
@mrlukasbos You are my hero! It finally seems to work! I can control all my Plejd-devices, their status are instantly updated in Homey and they have been stable for about 20 hours.

**Comment by mrlukasbos** [comment]
Good to hear. Please let us know if you experience any other issues.
